storage_controller/src/scheduler.rs

```diff
@@ -363,16 +363,20 @@ impl Scheduler {
                         *k,
                         context.nodes.get(k).copied().unwrap_or(AffinityScore::FREE),
                         v.shard_count,
+                        v.attached_shard_count,
                     ))
                 }
             })
             .collect();
 
         // Sort by, in order of precedence:
-        //  1st: Affinity score.  We should never pick a higher-score node if a lower-score node is available
-        //  2nd: Utilization.  Within nodes with the same affinity, use the least loaded nodes.
-        //  3rd: Node ID.  This is a convenience to make selection deterministic in tests and empty systems.
-        scores.sort_by_key(|i| (i.1, i.2, i.0));
+        //  1st: Attached shard count.  We should always pick the node with the least number of
+        //  attached shards.
+        //  2nd: Affinity score.  Within nodes with the same count of attached shards, prefer
```

Comment:
On the order of precedence: I think we can't put attached shard count ahead of affinity score, or if a node has a low attached count, we'll end up putting all the shards for a sharded tenant onto that one node.

We'll ultimately need to do something smarter here, but as an incremental change it might make sense to put the attached shard count immediately after the utilization score, so that we let attached count influence things if the affinity score and utilization (total shard count) are equal.  So this line would be:
```
scores.sort_by_key(|i| (i.1, i.2, i.3, i.0));
```

That way in a simple test environment, we'll get the behavior we want that if we have 2 pageservers and create 2 single sharded tenants, each pageserver gets one attached location each.

Reply:
Yes, realised this is incorrect after our discussion today. Will update.

Reply:
I agree affinity should be first, but if we add the attached shard count *after* the total shard count, I don't think it would change anything. I propose we go: affinity, attached shard count, total shard count, node id.

Reply:
ok - that's not objectively worse than the way we do it today.